src/NonLinearProgram/nlp_utilities.jl

```diff
@@ -0,0 +1,556 @@
+# Copyright (c) 2020: Andrew Rosemberg and contributors
+#
+# Use of this source code is governed by an MIT-style license that can be found
+# in the LICENSE.md file or at https://opensource.org/licenses/MIT.
+
+"""
+    fill_off_diagonal(H)
+
+Filling the off-diagonal elements of a sparse matrix to make it symmetric.
+"""
+function fill_off_diagonal(H)
+    ret = H + H'
+    row_vals = SparseArrays.rowvals(ret)
+    non_zeros = SparseArrays.nonzeros(ret)
+    for col in 1:size(ret, 2)
+        for i in SparseArrays.nzrange(ret, col)
+            if col == row_vals[i]
+                non_zeros[i] /= 2
+            end
+        end
+    end
+    return ret
+end
+
+"""
+    compute_optimal_hessian(evaluator::MOI.Nonlinear.Evaluator, rows::Vector{JuMP.ConstraintRef}, x::Vector{JuMP.VariableRef})
+
+Compute the optimal Hessian of the Lagrangian.
+"""
+function compute_optimal_hessian(
+    model::Model,
+    rows::Vector{MOI.Nonlinear.ConstraintIndex},
+)
+    sense_multiplier = sense_mult(model)
+    evaluator = model.cache.evaluator
+    y = [model.y[model.model.nlp_index_2_constraint[row].value] for row in rows]
+    hessian_sparsity = MOI.hessian_lagrangian_structure(evaluator)
+    I = [i for (i, _) in hessian_sparsity]
+    J = [j for (_, j) in hessian_sparsity]
+    V = zeros(length(hessian_sparsity))
+    # The signals are being sdjusted to match the Ipopt convention (inner.mult_g)
+    # but we don't know if we need to adjust the objective function multiplier
+    MOI.eval_hessian_lagrangian(
+        evaluator,
+        V,
+        model.x,
+        1.0,
+        -sense_multiplier * y,
+    )
+    num_vars = length(model.x)
+    H = SparseArrays.sparse(I, J, V, num_vars, num_vars)
+    return fill_off_diagonal(H)
+end
+
+"""
+    compute_optimal_jacobian(evaluator::MOI.Nonlinear.Evaluator, rows::Vector{JuMP.ConstraintRef}, x::Vector{JuMP.VariableRef})
+
+Compute the optimal Jacobian of the constraints.
+"""
+function compute_optimal_jacobian(
+    model::Model,
+    rows::Vector{MOI.Nonlinear.ConstraintIndex},
+)
+    evaluator = model.cache.evaluator
+    jacobian_sparsity = MOI.jacobian_structure(evaluator)
+    I = [i for (i, _) in jacobian_sparsity]
+    J = [j for (_, j) in jacobian_sparsity]
+    V = zeros(length(jacobian_sparsity))
+    MOI.eval_constraint_jacobian(evaluator, V, model.x)
+    A = SparseArrays.sparse(I, J, V, length(rows), length(model.x))
+    return A
+end
+
+"""
+    compute_optimal_hess_jac(evaluator::MOI.Nonlinear.Evaluator, rows::Vector{JuMP.ConstraintRef}, x::Vector{JuMP.VariableRef})
+
+Compute the optimal Hessian of the Lagrangian and Jacobian of the constraints.
```

Comment:
What is the "optimal` Hessian? Do you mean the Hessian evaluated at the solution?

Reply:
yes, will edit!